src/identity/IdentityProviderHttpHandler.ts

```diff
@@ -162,39 +187,79 @@ export class IdentityProviderHttpHandler extends HttpHandler {
         const errorMessage = createErrorMessage(error);
-        return await this.handleTemplateResponse(response, route.viewTemplate, { errorMessage, prefilled });
+        return {
+          result: { type: 'response', details: { errorMessage, prefilled }},
+          templateFiles: route.viewTemplates,
+        };
       }
+    }
 
-      if (result.type === 'complete') {
-        return await this.interactionCompleter.handleSafe({ ...result.details, request, response });
-      }
-      if (result.type === 'response' && route.responseTemplate) {
-        return await this.handleTemplateResponse(response, route.responseTemplate, result.details);
-      }
+    throw new BadRequestHttpError(`Unsupported request: ${operation.method} ${operation.target.path}`);
+  }
+
+  /**
+   * Creates a ResponseDescription based on the InteractionHandlerResult.
+   * This will either be a redirect if type is "complete" or a data stream if the type is "response".
+   */
+  private async handleInteractionResult(operation: Operation, request: HttpRequest, result: InteractionHandlerResult,
+    templateFiles: Record<string, string>): Promise<ResponseDescription> {
+    let responseDescription: ResponseDescription | undefined;
+
+    if (result.type === 'complete') {
+      // Create a redirect URL with the OIDC library
+      const location = await this.interactionCompleter.handleSafe({ ...result.details, request });
+      responseDescription = new RedirectResponseDescription(location);
+    } else {
+      // Convert the response object to a data stream
+      responseDescription = await this.handleResponseResult(result, templateFiles, operation);
     }
-    throw new BadRequestHttpError(`Unsupported request: ${request.method} ${request.url}`);
+
+    return responseDescription;
   }
 
-  private async handleTemplateResponse(response: HttpResponse, templateFile: string, contents?: NodeJS.Dict<any>):
-  Promise<void> {
-    await this.templateHandler.handleSafe({ response, templateFile, contents: contents ?? {}});
+  /**
+   * Converts an InteractionResponseResult to a ResponseDescription by first converting to a Representation
+   * and applying necessary conversions.
+   */
+  private async handleResponseResult(result: InteractionResponseResult, templateFiles: Record<string, string>,
+    operation: Operation): Promise<ResponseDescription> {
+    // Convert the object to a valid JSON representation
+    const json = { ...result.details, apiVersion: API_VERSION };
+    const representation = new BasicRepresentation(JSON.stringify(json), operation.target, APPLICATION_JSON);
+
+    // Template metadata is required for conversion
+    for (const [ type, templateFile ] of Object.entries(templateFiles)) {
+      const templateNode = namedNode(templateFile);
+      representation.metadata.add(SOLID_META.terms.template, templateNode);
+      representation.metadata.addQuad(templateNode, CONTENT_TYPE_TERM, type);
```

Comment:
We should definitely be able to do this without mentions of quads and named nodes.

Reply:
So a helper function somewhere else? That would make sense.

Reply:
Or the `metadata` object having helpers that take all strings.